File: GroupAnagrams.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string, vector<string>> groups;
        for (string str : strs) {
            int freq[26];
            memset(freq, 0, 26*sizeof(int));
            for (int i = 0; i < str.size(); i++)
                freq[str[i]-'a']++;
            string key;
            key.reserve(str.size());
            for (int i = 0; i < 26; i++) {
                char ch = i+'a';
                for (int j = 0; j < freq[i]; j++) {
                    key += ch;
                }
            }
            groups[key].push_back(str);
        }
        vector<vector<string>> ans;
        for (auto it : groups) {
            ans.push_back(it.second);
        }
        return ans;
    }
};

File: GroupAnagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        HashMap<String, List<String>> map = new HashMap<>();
        for (String s : strs) {
            String key = getKey(s);
            if (!map.containsKey(key)) {
                map.put(key, new ArrayList<>());
            }
            map.get(key).add(s);
        }
        List<List<String>> res = new ArrayList<>(map.keySet().size());
        for (Map.Entry<String, List<String>> entry : map.entrySet()) {
            res.add(entry.getValue());
        }
        return res;
    }

    private String getKey(String s) {
        int[] freq = new int[26];
        for (int i = 0; i < s.length(); ++i) {
            ++freq[s.charAt(i)-'a'];
        }
        char[] res = new char[s.length()];
        for (int i = 0, j = 0; i < 26; ++i) {
            while (freq[i] > 0) {
                res[j++] = (char)(i+'a');
                --freq[i];
            }
        }
        return new String(res);
    }
}
